libs/parallelism/executors/include/hpx/executors/scheduler_executor.hpp

```diff
@@ -0,0 +1,319 @@
+        constexpr bool operator==(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ == rhs.sched_;
+        }
+
+        constexpr bool operator!=(scheduler_executor const& rhs) const noexcept
+        {
+            return sched_ != rhs.sched_;
+        }
+
+        constexpr auto const& context() const noexcept
+        {
+            return *this;
+        }
+        /// \endcond
+
+        /// \cond NOINTERNAL
+        // Associate the parallel_execution_tag executor tag type as a default
+        // with this executor.
+        using execution_category = parallel_execution_tag;
+
+        // Associate the static_chunk_size executor parameters type as a default
+        // with this executor.
+        using executor_parameters_type = static_chunk_size;
+
+        template <typename T, typename... Ts>
+        using future_type = hpx::future<T>;
+
+        // NonBlockingOneWayExecutor interface
+        template <typename F, typename... Ts>
+        void post(F&& f, Ts&&... ts)
+        {
+            detach(transform(schedule(sched_),
+                detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // OneWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) sync_execute(F&& f, Ts&&... ts)
+        {
+            return sync_wait(transform(schedule(sched_),
+                detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // TwoWayExecutor interface
+        template <typename F, typename... Ts>
+        decltype(auto) async_execute(F&& f, Ts&&... ts)
+        {
+            return make_future(transform(schedule(sched_),
+                detail::captured_args(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        template <typename F, typename Future, typename... Ts>
+        decltype(auto) then_execute(F&& f, Future&& predecessor, Ts&&... ts)
+        {
+            auto&& sched =
+                on(keep_future(std::forward<Future>(predecessor)), sched_);
+
+            return make_future(transform(std::move(sched),
+                detail::captured_args_unary(
+                    std::forward<F>(f), std::forward<Ts>(ts)...)));
+        }
+
+        // BulkTwoWayExecutor interface
+        template <typename F, typename S, typename... Ts>
+        decltype(auto) bulk_async_execute(F&& f, S const& shape, Ts&&... ts)
+        {
+            using shape_element =
+                typename hpx::traits::range_traits<S>::value_type;
+            using result_type = hpx::util::detail::invoke_deferred_result_t<F,
+                shape_element, Ts...>;
+
+            std::vector<hpx::future<result_type>> results;
+            results.reserve(hpx::util::size(shape));
+
+            for (auto const& s : shape)
+            {
+                results.push_back(async_execute(f, s, ts...));
+            }
```

Comment:
Should this not also be based on `bulk`? Something in the direction of this:
```
return make_future(bulk(schedule(sched_), size(shape), something_to_bind_ts_and_possibly_a_vector_of_results);
```
Or if we're lazy:
```
return bulk_then_execute(std::forward<F>(f), shape, make_ready_future(), std::forward<Ts>(ts)...);
```
?
```

Reply:
Nod, I'll try that.

Reply:
The second option wouldn't really work as the return types of `bulk_async_execute` and `bulk_then_execute` are slightly different. `bulk_async_execute` returns a `vector<future<result_type>>`, while `bulk_then_execute` returns a `future<vector<result_type>>`. I will wait for `bulk` to support concurrent scheduling before changing the implementation here.

Reply:
> The second option wouldn't really work as the return types of bulk_async_execute and bulk_then_execute are slightly different. bulk_async_execute returns a vector<future<result_type>>, while bulk_then_execute returns a future<vector<result_type>>.
I missed that, good point.

> I will wait for bulk to support concurrent scheduling before changing the implementation here.

Does it make a difference? The implementation here should look the same before and after #5487.

Reply:
> Does it make a difference? The implementation here should look the same before and after #5487.

*grumble* I was hoping to get away with it,  but you caught me... :-)

Reply:
I don't mind doing it if you'd like :P I just wanted to understand if there was a technical reason for not changing it now.

Reply:
@msimberg I give up, I can't see how this could possibly work. I have removed most of the custom binders now, please feel free to take a stab at `bulk_async_execute` if you like.

Reply:
I'll have a look. Perhaps I was too optimistic and I'll get stuck as well...

Reply:
I've pushed an attempt at it, but I can see now why you didn't like it. I don't particularly like it either as it's far from a portable definition (this won't work with a GPU scheduler).

The problem is really the return type. If we were to accept the same relaxation `parallel_executor_aggregated` has on its return type from `bulk_async_execute` things would be significantly simpler, but it does of course lose the futures to the individual tasks (which is anyway something that mainly makes sense on the host, not on a GPU).

Reply:
As discussed, I think we can safely return a `std::vector<hpx::future<void>>`  with one element in case the return type is `void`. Otherwise, I think we should go with what you suggest. For the former, we might even get away with returning a `std::array<hpx::future<void>, 1>`, this might require (I'm not sure) slight changes to the algorithms, however.

Reply:
Updated. I did not make it a `std::array<hpx::future<void>, 1>` since that would have required larger changes to the algorithms.

As another flyby, I did what should've been done in #5455: remove the `as_sender` fallback for `schedule`. It turned out that the `scheduler_executor` was identifying as a scheduler, meaning that the one explicit constructor (https://github.com/STEllAR-GROUP/hpx/blob/c5465617297b1cf8189941c82d93e0c5b315f154/libs/parallelism/executors/include/hpx/executors/scheduler_executor.hpp#L70-L76) was being considered instead of the real copy constructor.